src/goto-instrument/contracts/contracts.cpp

```diff
@@ -140,123 +140,140 @@ void code_contractst::check_apply_loop_contracts(
-
-  // find the last back edge
-  goto_programt::targett loop_end = loop_head;
-  for(const auto &t : loop)
-    if(
-      t->is_goto() && t->get_target() == loop_head &&
-      t->location_number > loop_end->location_number)
-      loop_end = t;
-
-  // check for assigns, invariant, and decreases clauses
-  auto assigns_clause = static_cast<const exprt &>(
-    loop_end->get_condition().find(ID_C_spec_assigns));
-  auto invariant = static_cast<const exprt &>(
-    loop_end->get_condition().find(ID_C_spec_loop_invariant));
-  auto decreases_clause = static_cast<const exprt &>(
-    loop_end->get_condition().find(ID_C_spec_decreases));
-
-  if(invariant.is_nil())
-  {
-    if(decreases_clause.is_nil() && assigns_clause.is_nil())
-      return;
-    else
-    {
-      invariant = true_exprt();
-      log.warning()
-        << "The loop at " << loop_head->source_location().as_string()
-        << " does not have an invariant, but has other clauses"
-        << " specified in its contract.\n"
-        << "Hence, a default invariant ('true') is being used to prove those."
-        << messaget::eom;
-    }
-  }
-  else
-  {
-    // form the conjunction
-    invariant = conjunction(invariant.operands());
-  }
+  const auto loop_head_location = loop_head->source_location();
 
   // Vector representing a (possibly multidimensional) decreases clause
   const auto &decreases_clause_exprs = decreases_clause.operands();
 
   // Temporary variables for storing the multidimensional decreases clause
   // at the start of and end of a loop body
-  std::vector<symbol_exprt> old_decreases_vars;
-  std::vector<symbol_exprt> new_decreases_vars;
+  std::vector<symbol_exprt> old_decreases_vars, new_decreases_vars;
+
+  replace_symbolt replace;
+  code_contractst::add_quantified_variable(invariant, replace, mode);
+  replace(invariant);
 
-  // change
-  //   H: loop;
-  //   E: ...
+  // instrument
+  //
+  //   ... preamble ...
+  // HEAD:
+  //   ... eval guard ...
+  //   if (!guard)
+  //     goto EXIT;
+  //   ... loop body ...
+  //   goto HEAD;
+  // EXIT:
+  //   ... postamble ...
+  //
   // to
-  //   initialize loop_entry variables;
-  //   H: assert(invariant);
-  //   snapshot(write_set);
-  //   havoc;
-  //   assume(invariant);
-  //   if(guard) goto E:
-  //   old_decreases_value = decreases_clause_expr;
-  //   loop with optional write_set inclusion checks;
-  //   new_decreases_value = decreases_clause_expr;
-  //   assert(invariant);
-  //   assert(new_decreases_value < old_decreases_value);
-  //   assume(false);
-  //   E: ...
+  //
+  //                               ... preamble ...
+  //                        ,-     initialize loop_entry history vars;
+  //                        |      entered_loop = false
+  // loop assigns check     |      initial_invariant_val = invariant_expr;
+  //  - unchecked, temp     |      in_base_case = true;
+  // func assigns check     |      snapshot (write_set);
+  //  - disabled via pragma |      goto HEAD;
```

Comment:
disable what via pragma?

Reply:
`func assigns check` is disabled by pragma